### DISCUSSION TITLE
Error generating test sentences on multi-speaker configuration

### DISCUSSION THREAD

**Discussion by astricks** [coqui-ai/TTS#496] [General Q&A] [answered]
I'm trying to train a multi-speaker model of 7 speakers. However, test sentence generation fails with the following stacktrace:

-->8--
!! Error creating Test Sentence - 1                                                                                                                                          
Traceback (most recent call last):                                                                                                                                                
  File "/home/arvind/code/TTS/TTS/bin/train_tacotron.py", line 473, in evaluate                                                                                                   
    do_trim_silence=False)                                                                                                                                                        
  File "/home/arvind/code/TTS/TTS/tts/utils/synthesis.py", line 266, in synthesis                                                                                                 
    model, inputs, CONFIG, truncated, speaker_id, style_mel, speaker_embeddings=speaker_embedding)                                                                                
  File "/home/arvind/code/TTS/TTS/tts/utils/synthesis.py", line 68, in run_model_torch                                                                                            
    inputs, speaker_ids=speaker_id, speaker_embeddings=speaker_embeddings)                                                                                                        
  File "/home/arvind/code/TTS/venv/lib/python3.6/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context                                                         
    return func(*args, **kwargs)                                                                                                                                                  
  File "/home/arvind/code/TTS/TTS/tts/models/tacotron2.py", line 201, in inference                                                                                                
    encoder_outputs = self._concat_speaker_embedding(encoder_outputs, speaker_embeddings)                                                                                         
  File "/home/arvind/code/TTS/TTS/tts/models/tacotron_abstract.py", line 223, in _concat_speaker_embedding                                                                        
    outputs.size(0), outputs.size(1), -1)
RuntimeError: The expanded size of the tensor (77) must match the existing size (512) at non-singleton dimension 1.  Target sizes: [1, 77, -1].  Tensor sizes: [512, 1]           
-->8--

The single speaker training worked just fine, and I got fairly good results. My config is below. The only change from the single speaker config I made is to change use_speaker_embedding to true and external_speaker_embedding_file to null.

{
    "model": "Tacotron2",
    "run_name": "hindirun",
    "run_description": "hindirun without gst",

    // AUDIO PARAMETERS
    "audio":{
        // stft parameters
        "fft_size": 1024,         // number of stft frequency levels. Size of the linear spectogram frame.
        "win_length": 1024,      // stft window length in ms.
        "hop_length": 256,       // stft window hop-lengh in ms.
        "frame_length_ms": null, // stft window length in ms.If null, 'win_length' is used.
        "frame_shift_ms": null,  // stft window hop-lengh in ms. If null, 'hop_length' is used.

        // Audio processing parameters
        "sample_rate": 22050,   // DATASET-RELATED: wav sample-rate.
        "preemphasis": 0.0,     // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.
        "ref_level_db": 20,     // reference level db, theoretically 20db is the sound of air.

        // Silence trimming
        "do_trim_silence": true,// enable trimming of slience of audio as you load it. LJspeech (true), TWEB (false), Nancy (true)
        "trim_db": 40,          // threshold for timming silence. Set this according to your dataset.

        // Griffin-Lim
        "power": 1.5,           // value to sharpen wav signals after GL algorithm.
        "griffin_lim_iters": 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.

        // MelSpectrogram parameters
        "num_mels": 80,         // size of the mel spec frame.
        "mel_fmin": 50.0,        // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!
        "mel_fmax": 10000.0,     // maximum freq level for mel-spec. Tune for dataset!!
        "spec_gain": 1,

        // Normalization parameters
        "signal_norm": true,    // normalize spec values. Mean-Var normalization if 'stats_path' is defined otherwise range normalization defined by the other params.
        "min_level_db": -100,   // lower bound for normalization
        "symmetric_norm": true, // move normalization to range [-1, 1]
        "max_norm": 4.0,        // scale normalization to range [-max_norm, max_norm] or [0, max_norm]
        "clip_norm": true,      // clip normalized values into the range.
        "stats_path": null      // DO NOT USE WITH MULTI_SPEAKER MODEL. scaler stats file computed by 'compute_statistics.py'. If it is defined, mean-std based notmalization is used and other normalization params are ignored
    },

    // VOCABULARY PARAMETERS
    // if custom character set is not defined,
    // default set in symbols.py is used
    "characters":{
        "pad": "_",
        "eos": "~",
        "bos": "^",
        "characters": "अआइईउऊऋएऐऑओऔकखगघङचछजझञटठडढणतथदधनपफबभमयरलवशषसहह़ािीुूृेैॉोौ्",
        "punctuations":"!,-.? ",
        "phonemes":"iyɨʉɯuɪʏʊeøɘəɵɤoɛœɜɞʌɔæɐaɶɑɒᵻʘɓǀɗǃʄǂɠǁʛpbtdʈɖcɟkɡqɢʔɴŋɲɳnɱmʙrʀⱱɾɽɸβfvθðszʃʒʂʐçʝxɣχʁħʕhɦɬɮʋɹɻjɰlɭʎʟˈˌːˑʍwɥʜʢʡɕʑɺɧɚ˞ɫ"
    },

    // DISTRIBUTED TRAINING
    "distributed":{
        "backend": "nccl",
        "url": "tcp:\/\/localhost:54321"
    },

    "reinit_layers": [],    // give a list of layer names to restore from the given checkpoint. If not defined, it reloads all heuristically matching layers.

    // TRAINING
    "batch_size": 32,       // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.
    "eval_batch_size":16,
    "r": 8,                 // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled.
    "gradual_training": null, 
    "mixed_precision": false,     // level of optimization with NVIDIA's apex feature for automatic mixed FP16/FP32 precision (AMP), NOTE: currently only O1 is supported, and use "O1" to activate.

    // LOSS SETTINGS
    "loss_masking": true,       // enable / disable loss masking against the sequence padding.
    "decoder_loss_alpha": 0.5,  // original decoder loss weight. If > 0, it is enabled
    "postnet_loss_alpha": 0.25, // original postnet loss weight. If > 0, it is enabled
    "postnet_diff_spec_alpha": 0.25,     // differential spectral loss weight. If > 0, it is enabled
    "decoder_diff_spec_alpha": 0.25,     // differential spectral loss weight. If > 0, it is enabled
    "decoder_ssim_alpha": 0.5,     // decoder ssim loss weight. If > 0, it is enabled
    "postnet_ssim_alpha": 0.25,     // postnet ssim loss weight. If > 0, it is enabled
    "ga_alpha": 5.0,           // weight for guided attention loss. If > 0, guided attention is enabled.
    "stopnet_pos_weight": 15.0, // pos class weight for stopnet loss since there are way more negative samples than positive samples.

    // VALIDATION
    "run_eval": true,
    "test_delay_epochs": 10,  //Until attention is aligned, testing only wastes computation time.
    "test_sentences_file": "/home/arvind/data/hindirun/metadata_test_sentences.txt",  // set a file to load sentences to be used for testing. If it is null then we use default english sentences.

    // OPTIMIZER
    "noam_schedule": false,        // use noam warmup and lr schedule.
    "grad_clip": 1.0,              // upper limit for gradients for clipping.
    "epochs": 20000,               // total number of epochs to train.
    "lr": 0.0001,                  // Initial learning rate. If Noam decay is active, maximum learning rate.
    "wd": 0.000001,                // Weight decay weight.
    "warmup_steps": 4000,          // Noam decay steps to increase the learning rate from 0 to "lr"
    "seq_len_norm": false,         // Normalize eash sample loss with its length to alleviate imbalanced datasets. Use it if your dataset is small or has skewed distribution of sequence lengths.

    // TACOTRON PRENET
    "memory_size": -1,             // ONLY TACOTRON - size of the memory queue used fro storing last decoder predictions for auto-regression. If < 0, memory queue is disabled and decoder only uses the last prediction frame.
    "prenet_type": "original",     // "original" or "bn".
    "prenet_dropout": false,       // enable/disable dropout at prenet.

    // TACOTRON ATTENTION
    "attention_type": "original",  // 'original' , 'graves', 'dynamic_convolution'
    "attention_heads": 4,          // number of attention heads (only for 'graves')
    "attention_norm": "sigmoid",   // softmax or sigmoid.
    "windowing": false,            // Enables attention windowing. Used only in eval mode.
    "use_forward_attn": false,     // if it uses forward attention. In general, it aligns faster.
    "forward_attn_mask": false,    // Additional masking forcing monotonicity only in eval mode.
    "transition_agent": false,     // enable/disable transition agent of forward attention.
    "location_attn": true,         // enable_disable location sensitive attention. It is enabled for TACOTRON by default.
    "bidirectional_decoder": false,  // use https://arxiv.org/abs/1907.09006. Use it, if attention does not work well with your dataset.
    "double_decoder_consistency": false,  // use DDC explained here https://erogol.com/solving-attention-problems-of-tts-models-with-double-decoder-consistency-draft/
    "ddc_r": 4,                           // reduction rate for coarse decoder.

    // STOPNET
    "stopnet": true,               // Train stopnet predicting the end of synthesis.
    "separate_stopnet": true,      // Train stopnet seperately if 'stopnet==true'. It prevents stopnet loss to influence the rest of the model. It causes a better model, but it trains SLOWER.

    // TENSORBOARD and LOGGING
    "print_step": 25,       // Number of steps to log training on console.
    "tb_plot_step": 100,    // Number of steps to plot TB training figures.
    "print_eval": true,     // If True, it prints intermediate loss values in evalulation.
    "save_step": 5000,      // Number of training steps expected to save traninpg stats and checkpoints.
    "checkpoint": true,     // If true, it saves checkpoints per "save_step"
    "keep_all_best": false,  // If true, keeps all best_models after keep_after steps
    "keep_after": 10000,    // Global step after which to keep best models if keep_all_best is true
    "tb_model_param_stats": false,     // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.

    // DATA LOADING
    "text_cleaner": "basic_cleaners",
    "enable_eos_bos_chars": false, // enable/disable beginning of sentence and end of sentence chars.
    "num_loader_workers": 2,        // number of training data loader processes. Don't set it too big. 4-8 are good values.
    "num_val_loader_workers": 2,    // number of evaluation data loader processes.
    "batch_group_size": 32,  //Number of batches to shuffle after bucketing.
    "min_seq_len": 24,      // DATASET-RELATED: minimum text length to use in training
    "max_seq_len": 130,     // DATASET-RELATED: maximum text length
    "compute_input_seq_cache": false,  // if true, text sequences are computed before starting training. If phonemes are enabled, they are also computed at this stage.
    "use_noise_augment": true,

    // PATHS
    "output_path": "/home/arvind/code/output",

    // PHONEMES
    "phoneme_cache_path": "/home/arvind/code/cache/",  // phoneme computation is slow, therefore, it caches results in the given folder.
    "use_phonemes": false,           // use phonemes instead of raw characters. It is suggested for better pronounciation.
    "phoneme_language": "hi",     // depending on your target language, pick one from  https://github.com/bootphon/phonemizer#languages

    // MULTI-SPEAKER and GST
    "use_speaker_embedding": true,      // use speaker embedding to enable multi-speaker learning.
    "use_gst": false,          	        // use global style tokens
    "use_external_speaker_embedding_file": false, // if true, forces the model to use external embedding per sample instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "external_speaker_embedding_file": null, // if not null and use_external_speaker_embedding_file is true, it is used to load a specific embedding file and thus uses these embeddings instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "gst":	{			                // gst parameter if gst is enabled
        "gst_style_input": null,        // Condition the style input either on a
                                        // -> wave file [path to wave] or
                                        // -> dictionary using the style tokens {'token1': 'value', 'token2': 'value'} example {"0": 0.15, "1": 0.15, "5": -0.15}
                                        // with the dictionary being len(dict) <= len(gst_style_tokens).
        "gst_embedding_dim": 512,
        "gst_num_heads": 4,
        "gst_style_tokens": 10,
        "gst_use_speaker_embedding": false
	},

    // DATASETS
    "datasets":   // List of datasets. They all merged and they get different speaker_ids.
        [
            {
                "name": "hindirun",
                "path": "/home/arvind/data/hindirun",
                "meta_file_train": "metadata_train.txt", // for vtck if list, ignore speakers id in list for train, its useful for test cloning with new speakers
                "meta_file_val": null 
            }
        ]
}
--

I've been trying to debug this to no avail. What am I missing?

**Comment by erogol** [comment]
is it possible for you to run this on colab and share it. So I can check what is missing.

**Reply by astricks** [reply]
I'll try spinning up a colab

**Reply by astricks** [reply]
@erogol I've spun up a colab, not sure how to share the data though since it references my Google drive.
https://colab.research.google.com/drive/1E7MzM0klciW7sONN59N4uhhgtQwdLBpR?usp=sharing
Any suggestions? The code is all there, it just needs a multispeaker dataset.

**Reply by astricks** [reply]
I'm not sure when the variable speaker_embedding should even be initialized, that's what is confusing me. Its None, which is the issue I'm crashing on.

**Reply by erogol** [reply]
will look at it tomorrow.

**Reply by astricks** [reply]
I'd appreciate any help I can get on this - I'm not able to figure out how to run a multispeaker model, have it generate the test sentences, and do inference on the trained model.

**Reply by erogol** [reply]
colab relies on your gdrive to fetch the data I guess. Thus I couldn't run the code.

**Reply by astricks** [reply]
Yeah, I figured. No worries.

**Comment by lexkoro** [comment, marked as answer]
I think the dimensions are wrong during inference.

Can you try adding this line to your code?
https://github.com/SanjaESC/TTS-test/blob/94c51eda1d98c8dfb7341972c070937c7ec1f4d2/TTS/tts/models/tacotron2.py#L258

Tests are run on external embeddings only I guess thus no errors there?

**Reply by astricks** [reply]
This worked, thanks @SanjaESC!
I was trying to put in a fix at this spot but was doing it incorrectly.

**Reply by astricks** [reply]
With test_sentences now being generated, I tried to generate some speech using the model. Here I hit my next issue. It seems I need to tell it somehow that the model has speaker_embedding enabled. Not sure what I'm overlooking.

--
> python3 TTS/bin/synthesize.py --text "मैं बता रहा हूँ, मुझे यहाँ कुछ ठीक नहीं लग रहा, यार।" --model_path ../output/kamran-May-28-2021_01+52PM-56ae83a/best_model.pth.tar --config_path ../output/kamran-May-28-2021_01+52PM-56ae83a/config.json --out_path /tmp/wavernn4.wav --speakers_file_path ../output/kamran-May-28-2021_01+52PM-56ae83a/speakers.json --speaker_idx 0
Loading speakers ...
 > Using model: Tacotron2
Traceback (most recent call last):
  File "TTS/bin/synthesize.py", line 265, in <module>
    main()
  File "TTS/bin/synthesize.py", line 234, in main
    args.use_cuda,
  File "/home/arvind/code/TTS/TTS/utils/synthesizer.py", line 72, in __init__
    self._load_tts(tts_checkpoint, tts_config_path, use_cuda)
  File "/home/arvind/code/TTS/TTS/utils/synthesizer.py", line 142, in _load_tts
    self.tts_model.load_checkpoint(self.tts_config, tts_checkpoint, eval=True)
  File "/home/arvind/code/TTS/TTS/tts/models/tacotron_abstract.py", line 132, in load_checkpoint
    self.load_state_dict(state["model"])
  File "/home/arvind/code/TTS/venv0013/lib/python3.6/site-packages/torch/nn/modules/module.py", line 1224, in load_state_dict
    self.__class__.__name__, "\n\t".join(error_msgs)))
RuntimeError: Error(s) in loading state_dict for Tacotron2:
	Unexpected key(s) in state_dict: "speaker_embedding.weight". 
	size mismatch for decoder.attention_rnn.weight_ih: copying a param with shape torch.Size([4096, 1280]) from checkpoint, the shape in current model is torch.Size([4096, 768]).
	size mismatch for decoder.attention.inputs_layer.linear_layer.weight: copying a param with shape torch.Size([128, 1024]) from checkpoint, the shape in current model is torch.Size([128, 512]).
	size mismatch for decoder.decoder_rnn.weight_ih: copying a param with shape torch.Size([4096, 2048]) from checkpoint, the shape in current model is torch.Size([4096, 1536]).
	size mismatch for decoder.linear_projection.linear_layer.weight: copying a param with shape torch.Size([640, 2048]) from checkpoint, the shape in current model is torch.Size([640, 1536]).

**Reply by erogol** [reply]
@SanjaESC why not PR?

**Reply by lexkoro** [reply]
I didn't test it, but if it fixes the problem ill do a PR

**Reply by astricks** [reply]
Should I mark this thread as answered and start a new discussion with the issue above?

**Reply by astricks** [reply]
Marking this as answered, moving the second question to another discussion.
